### PR TITLE
ci: prune unpushed images before acquisition

### LINKED ISSUE #2695: CI takes much too long
CI builds are regularly taking 20m+. This is much too long.

My ideal would be <10m, but that is unrealistic with the Rust toolchain. At the very least, we should strive to keep all test jobs <3m. This is true of all jobs except for testdrive, which takes now takes ~7m. We should investigate whether that is caused by slowness on our end, and, if not, look into sharding testdrive runs into multiple jobs.

### PR DESCRIPTION
**Pull request by benesch**
@petrosagg sending this one your way since I know you've poked at the build job in CI before.

----

This fixes one of my longest standing gripes with CI, which is that when
all images already exist (e.g. because no code has changed), the build
job still has to download all the existing images from Docker Hub.

This commit reworks the build job so that the only images downloaded are
those that are dependencies of other images that need to be built.

Fix #2695.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Review by petrosagg — APPROVED** [review]
Amazing, CI passed in 9 minutes! It clearly seems to be working as advertised but I'm not sure I understand what's going on.

Are the images we're skipping build-dependencies only and we're not using them for running the tests themselves? If yes, why are we even attempting to prefetch them and not just pull them when an individual mzbuild actually runs. Afaik docker will automatically consolidate concurrent pulls for the same spec and will also skip re-downloading if the registry manifest matches the local one

**Comment by benesch** [comment]
> Amazing, CI passed in 9 minutes! It clearly seems to be working as advertised but I'm not sure I understand what's going on.

Yeah unfortunately it won't be quite so smooth if you actually, you know, change any code, but it is nice in the cases when you don't!

> Are the images we're skipping build-dependencies only and we're not using them for running the tests themselves? If yes, why are we even attempting to prefetch them and not just pull them when an individual mzbuild actually runs. Afaik docker will automatically consolidate concurrent pulls for the same spec and will also skip re-downloading if the registry manifest matches the local one

Ah, ok. The code here only impacts the "build" job, whose responsibility in life is to make sure that every image in the entire repository is available on Docker Hub. Then when Buildkite farms out the tests to agents, those agents only download the specific images they need for the test from Docker Hub.

The build job used to do this, roughly:

```python
for image in repo:
    image.acquire()
    image.push()
````

"acquire" means pull it from Docker Hub if you can, otherwise build it locally. This works just fine if you need to rebuild all the images. But suppose you've only made changes to `src/testdrive/src/lib.rs`. Then the only image that needs to be rebuilt is `testdrive`, but the old build job would waste a bunch of time "acquiring" (i.e., running `docker pull`) on `materialized`, `chbenchmark`, and all the other random images in the repo.

Morally, the new code skips acquiring any image that already exists on Docker Hub. But it's not quite so simple in the case of image dependencies, where you need to acquire an image that already exists on Docker Hub in order to build an image that depends upon that one ... you know, maybe I'm crazy, and it _is_ that simple. One sec.

**Comment by benesch** [comment]
Ok, right, unpublished images. I'm pretty sure I did get it right last night. Let's say you have this stack of dependencies:

```
ubuntu base <-- testdrive
     ^
     |
materialized
```

Let's say "ubuntu base" and "testdrive" haven't been changed (i.e., already exist on Docker Hub) but "materialized" has. The new code will skip acquiring "testdrive" because it's already on Docker Hub, but it will continue to acquire "ubuntu base" because "ubuntu base" is required to be available to build "materialized". If neither "materialized" nor "testdrive" has changed, the new code will skip acquiring everything, because it doesn't need "ubuntu base" to be available as a build dependency.

**Comment by petrosagg** [comment]
Oh, yeah I wasn't questioning if the code results in the logic you described. But imagine you had a function called `def ensure_exists_in_dockerhub(image)` that checked if the image exists in dockerhub and if not it would pull dependencies, build, and push.

In your example when neither `testdrive` nor `materialized` had changed, nothing gets pulled down.

If only `testdrive` is missing then by downloading the deps `ubuntu base` will be pulled.

If both `testdrive` and `materialized` are missing then the two build jobs will issue a `docker pull` for `ubuntu base`, but docker will ensure it happens only once.

So basically everything behaves dumb, but the smart thing ends up happening because of memoization

**Comment by benesch** [comment]
@petrosagg and I discussed offline for a bit. I think Petros's suggested approach is totally valid. I'm going to merge this as is and then do the refactor as a separate PR to see where it gets us.
